primitives/state-machine/src/ext.rs

```diff
@@ -757,6 +768,13 @@ where
 			(all_deleted, num_deleted)
 		} else {
 			self.backend.apply_to_keys_while(child_info, prefix, |key| {
+				if let Some(None) = match child_info {
```

Comment:
The fix here requires that no chain did call `clear_prefix_v2` twice in a block in the past, otherwise syncing will break. I cannot really say (for parity chains it was probably never the case).
To be safe we could also just keep the old code by applying this change only for `clear_prefix_v3`.
For instance replacing `limit` parameter to an enum with version could work (in a sense the limit defined at None is the old V1).

Reply:
Just though of another similar case: keyset { aa ab abc } is in state, then somewhere in block key ab is deleted then `clear_prefix_v2` is called with limit `2` : then with old implementation we have aa and ab removed and with new implementation the three key, leading to possible state root mismatch.